### ISSUE TITLE
Actor/Learner DQN Pong: Using PyUniformReplayBuffer instead of ReverbReplayBuffer not possible

### ISSUE THREAD

**Issue by Sch-Stef** [tensorflow/agents#797]
Hi together,

I try to adapt this [Atari Actor/Learner](https://github.com/tensorflow/agents/blob/635af30a6365ec963823d4623f14df25fcefaadd/tf_agents/experimental/examples/dqn/mnih15/dqn_train_eval_atari.py) example to my windows machine using a PyUniformReplayBuffer instead of the ReverbReplayBuffer.

I already used tf-agents DQN but with drivers (and the TFUniformReplayBuffer) instead of the (very elegant) Learner/Actor setup and it worked quite well.

The following adjustments were made in the code (marked separatly in code/snippet) :
- Use of **rb_observer = lambda x: replay_buffer.add_batch(batch_nested_array(x))**  --> Due to the (tf-agents internal) usage of the add_batch method (according to suggestions made by [issue post](https://github.com/tensorflow/agents/issues/198)
- Replaced **ReverbReplayBuffer with PyUniformReplayBuffer:** Added py_collect_data_spec = tensor_spec.to_array_spec(agent.collect_data_spec) in order to get the correct spec. 
- (Due to unfixable error "too many values to unpack (expected 2)") In Learner.py: Replaced "batched_specs, _ = dataset.element_spec" with "batched_specs = dataset.element_spec" in condition "if run_optimizer_variable_init"

After this I run in a few other errors that all contain data structure issues.
```
All of the Tensors in `value` must have two outer dimensions. Specifically, tensors must have shape `[B, T] + spec.shape`.
Full shapes of value tensors:
  Trajectory(
{'action': TensorShape([32]),
 'discount': TensorShape([32]),
 'next_step_type': TensorShape([32]),
 'observation': TensorShape([32, 84, 84, 4]),
 'policy_info': (),
 'reward': TensorShape([32]),
 'step_type': TensorShape([32])}).
Expected shapes (excluding the two outer dimensions):
  Trajectory(
{'action': TensorShape([]),
 'discount': TensorShape([]),
 'next_step_type': TensorShape([]),
 'observation': TensorShape([84, 84, 4]),
 'policy_info': (),
 'reward': TensorShape([]),
 'step_type': TensorShape([])}).
```

**--> I don't quite understand this, because the data should be created & exchanged according to the specifications resulting from the loading of the Atari environment. Therefore these errors should not occur - or is this use case simply not supported?**

I would be very grateful if someone could explain to me what I am missing or where my problem lies.

My code is as follows:

- Full:
[Ipynb_+_Pip_List.zip](https://github.com/tensorflow/agents/files/10059694/Ipynb_%2B_Pip_List.zip)

- Snippet:
```
def train_eval(
    root_dir,
    # env_name='Pong-v0',
    # env_name='PongNoFrameskip-v4',
    env_name='ALE/Pong-v5',
    # Training params
    update_frequency=4,  # Number of collect steps per policy update
    initial_collect_steps=50000,  # 50k collect steps
    num_iterations=1000000,  # 50M collect steps
    # Taken from Rainbow as it's not specified in Mnih,15.
    max_episode_frames_collect=50000,  # env frames observed by the agent
    max_episode_frames_eval=10000,  # env frames observed by the agent
    # Agent params
    epsilon_greedy=0.1,
    epsilon_decay_period=250000,  # 1M collect steps / update_frequency
    batch_size=32,
    learning_rate=0.00025,
    n_step_update=1,
    gamma=0.99,
    target_update_tau=1.0,
    target_update_period=2500,  # 10k collect steps / update_frequency
    reward_scale_factor=1.0,
    # Replay params
    # reverb_port=None,
    replay_capacity=1000,
    # Others
    policy_save_interval=250000,
    eval_interval=1000,
    eval_episodes=30,
    debug_summaries=True):
  """Trains and evaluates DQN."""

  collect_env = suite_atari.load(
      env_name,
      max_episode_steps=max_episode_frames_collect,
      gym_env_wrappers=suite_atari.DEFAULT_ATARI_GYM_WRAPPERS_WITH_STACKING)
  eval_env = suite_atari.load(
      env_name,
      max_episode_steps=max_episode_frames_eval,
      gym_env_wrappers=suite_atari.DEFAULT_ATARI_GYM_WRAPPERS_WITH_STACKING)

  observation_tensor_spec, action_tensor_spec, time_step_tensor_spec = (
      spec_utils.get_tensor_specs(collect_env))

  train_step = train_utils.create_train_step()

  num_actions = action_tensor_spec.maximum - action_tensor_spec.minimum + 1
  epsilon = tf.compat.v1.train.polynomial_decay(
      1.0,
      train_step,
      epsilon_decay_period,
      end_learning_rate=epsilon_greedy)

  agent = dqn_agent.DqnAgent(
      time_step_spec=time_step_tensor_spec,
      action_spec= action_tensor_spec,
      q_network=create_DL_q_network(num_actions),
      epsilon_greedy=epsilon,
      n_step_update=n_step_update,
      target_update_tau=target_update_tau,
      target_update_period=target_update_period,
      optimizer=tf.compat.v1.train.RMSPropOptimizer(
          learning_rate=learning_rate,
          decay=0.95,
          momentum=0.95,
          epsilon=0.01,
          centered=True),
      td_errors_loss_fn=common.element_wise_huber_loss,
      gamma=gamma,
      reward_scale_factor=reward_scale_factor,
      train_step_counter=train_step,
      debug_summaries=debug_summaries)
  
  ################ PyUniformReplayBuffer - START #########################
  py_collect_data_spec = tensor_spec.to_array_spec(agent.collect_data_spec)
  print("py_collect_data_spec : ",py_collect_data_spec)
  replay_buffer = py_uniform_replay_buffer.PyUniformReplayBuffer(
    data_spec= py_collect_data_spec,
    capacity=replay_capacity*batch_size
    )

  # observer of replay buffer
  #   rb_observer = replay_buffer.add_batch

  #********* Adpations add_batch method - START *********#
  rb_observer = lambda x: replay_buffer.add_batch(batch_nested_array(x))
  #********* Adpations add_batch method - END    *********#


  # create batch dataset
  dataset = replay_buffer.as_dataset( 
      sample_batch_size=32,
    #   num_steps = 1,
    #   num_steps = 2,
      single_deterministic_pass=False)
  
  experience_dataset_fn = lambda: dataset
  ################ PyUniformReplayBuffer - END #########################
  
  saved_model_dir = os.path.join(root_dir, learner.POLICY_SAVED_MODEL_DIR)
  env_step_metric = py_metrics.EnvironmentSteps()

# Create cyclic model saving trigger
  learning_triggers = [
      triggers.PolicySavedModelTrigger(
          saved_model_dir,
          agent,
          train_step,
          interval=policy_save_interval,
          metadata_metrics={triggers.ENV_STEP_METADATA_KEY: env_step_metric}),
      triggers.StepPerSecondLogTrigger(train_step, interval=100),
  ]

  dqn_learner = learner.Learner(
      root_dir,
      train_step,
      agent,
      experience_dataset_fn,
      triggers=learning_triggers,
      )

  # If we haven't trained yet make sure we collect some random samples first to
  # fill up the Replay Buffer with some experience.
  random_policy = random_py_policy.RandomPyPolicy(collect_env.time_step_spec(),
                                                  collect_env.action_spec())
  
  # tf_collect_env = TFPyEnvironment(collect_env)
  init_buffer_actor = actor.Actor(
      collect_env,
      random_policy,
      train_step,
      steps_per_run=initial_collect_steps,
      observers=[replay_buffer.add_batch, ShowProgress(initial_collect_steps)])
  
  logging.info('Doing initial collect.')
  print('Doing initial collect.')
  init_buffer_actor.run()

  tf_collect_policy = agent.collect_policy

  collect_policy = py_tf_eager_policy.PyTFEagerPolicy(tf_collect_policy,batch_time_steps=True,
                                                      use_tf_function=True)
                                                      
  collect_actor = actor.Actor(
      collect_env,
      collect_policy,
      train_step,
      steps_per_run=update_frequency,
      observers=[rb_observer, env_step_metric],
      metrics=actor.collect_metrics(10),
      summary_dir=os.path.join(root_dir, learner.TRAIN_DIR),
  )

  tf_greedy_policy = agent.policy
  greedy_policy = py_tf_eager_policy.PyTFEagerPolicy(tf_greedy_policy,batch_time_steps=True,
                                                     use_tf_function=True)

  eval_actor = actor.Actor(
      eval_env,
      greedy_policy,
      train_step,
      episodes_per_run=eval_episodes,
      metrics=actor.eval_metrics(eval_episodes),
      summary_dir=os.path.join(root_dir, 'eval'),
  )

  if eval_interval:
    logging.info('Evaluating.')
    print('Evaluating.')
    eval_actor.run_and_log()

  logging.info('Training.')
  print('Training.')
  for _ in range(num_iterations):
    collect_actor.run()
    dqn_learner.run(iterations=1)

    if eval_interval and dqn_learner.train_step_numpy % eval_interval == 0:
      logging.info('Evaluating.')
      eval_actor.run_and_log()
```

**Comment by coreyleveen** [comment]
I suggest trying the following:

1. Replacing the `PyUniformReplayBuffer` with a `TFUniformReplayBuffer`. It seems to conform better with the Learner API, and could address the `too many values to unpack` issue you described. For example:
```python
replay_buffer = tf_uniform_replay_buffer.TFUniformReplayBuffer(
  agent.collect_data_spec,
  batch_size,
  max_length=10000)
```
2. Uncommenting the `num_steps=2` line when creating the dataset. I believe this is the cause of the outer dimension issue.

**Comment by Sch-Stef** [comment]
Hi Corey,

thx for your feedback. I tried your modifications and after another small change it worked!

> 1. Replacing the `PyUniformReplayBuffer` with a `TFUniformReplayBuffer`. It seems to conform better with the Learner API, and could address the `too many values to unpack` issue you described. For example:
> 
> ```python
> replay_buffer = tf_uniform_replay_buffer.TFUniformReplayBuffer(
>   agent.collect_data_spec,
>   batch_size,
>   max_length=10000)
> ```

This "batch size" seems to be the main difference between the TFUniformReplayBuffer and the PyUniformReplayBuffer. Setting it to "1" when using the TFUniformReplayBuffer and using the "add_batch(batch_nested_array(x))" for the observer finally fixed it.

```
  # create buffer for storing experience
  replay_buffer = tf_uniform_replay_buffer.TFUniformReplayBuffer(
    agent.collect_data_spec,
    1,
    max_length=1000000)

 # create batch dataset
  dataset = replay_buffer.as_dataset( 
      sample_batch_size=32,
      num_steps = 2,
      single_deterministic_pass=False).prefetch(4)

  # create batched nested array input for rb_observer
  rb_observer = lambda x: replay_buffer.add_batch(batch_nested_array(x))

  # create batched readout of dataset
  experience_dataset_fn = lambda: dataset
```

So my explanation is as follows:
- The DQN saves steps in the Replay Buffer (each step has size one but contains 4 pics of 84x84)
- So the Batch Size of the Buffer needs to be "1" (I think I had tunnel vision... now that I read this, it's clear.)
- Long story short, the buffer size of the (exported) dataset needs to be 32.

Your assumption seems to be valid: The PyUniformReplayBuffer may not be fully suitable for the Actor Learner setup
-> I thought that these differences were handled by the "data-spec", which is supposed to be a blueprint for the buffer and determine how the data is processed....

Again thx for the help!
